File: ui/actionrender/colors.go
package actionrender

import (
	"image/color"
	"strings"

	"fyne.io/fyne/v2"
	"fyne.io/fyne/v2/theme"
)

func ActionCategoryForType(actionType string) string {
	switch strings.ToLower(strings.TrimSpace(actionType)) {
	case "move", "click", "key", "type":
		return "Mouse & Keyboard"
	case "imagesearch", "ocr", "findpixel":
		return "Detection"
	case "setvariable", "calculate", "datalist", "savevariable":
		return "Variables"
	case "wait", "focuswindow", "runmacro", "loop":
		return "Miscellaneous"
	default:
		return ""
	}
}

func ActionPastelColor(actionType string) color.NRGBA {
	t := strings.ToLower(strings.TrimSpace(actionType))
	category := ActionCategoryForType(t)
	isWait := t == "wait"

	isDark := fyne.CurrentApp().Settings().ThemeVariant() == theme.VariantDark
	if isDark {
		if isWait {
			return color.NRGBA{R: 0x7B, G: 0x4E, B: 0x3E, A: 0xFF}
		}
		switch category {
		case "Mouse & Keyboard":
			return color.NRGBA{R: 0x5E, G: 0x6B, B: 0x4A, A: 0xFF}
		case "Detection":
			return color.NRGBA{R: 0x5A, G: 0x4A, B: 0x44, A: 0xFF}
		case "Variables":
			return color.NRGBA{R: 0x7A, G: 0x63, B: 0x45, A: 0xFF}
		case "Miscellaneous":
			return color.NRGBA{R: 0x6A, G: 0x5A, B: 0x3F, A: 0xFF}
		default:
			return color.NRGBA{R: 0x5C, G: 0x54, B: 0x49, A: 0xFF}
		}
	}
	if isWait {
		return color.NRGBA{R: 0xC9, G: 0x8D, B: 0x6A, A: 0xFF}
	}
	switch category {
	case "Mouse & Keyboard":
		return color.NRGBA{R: 0xA1, G: 0xB0, B: 0x7A, A: 0xFF}
	case "Detection":
		return color.NRGBA{R: 0xB4, G: 0x9A, B: 0x84, A: 0xFF}
	case "Variables":
		return color.NRGBA{R: 0xC7, G: 0xAE, B: 0x7B, A: 0xFF}
	case "Miscellaneous":
		return color.NRGBA{R: 0xB8, G: 0x9A, B: 0x6A, A: 0xFF}
	default:
		return color.NRGBA{R: 0xB2, G: 0xA4, B: 0x8E, A: 0xFF}
	}
}

File: crates/sqyre-ui-model/src/colors.rs
//! Action tree color keys, pastels, and custom overrides.
//!
//! Hex parse/format helpers and category keys live in [`sqyre_domain::color`]
//! so vision and persist can use them without depending on UI chrome.

use sqyre_domain::{
    action_color_key as taxonomy_color_key, ACTION_COLOR_KEY_CONTROL_FLOW,
    ACTION_COLOR_KEY_DETECTION, ACTION_COLOR_KEY_MISCELLANEOUS, ACTION_COLOR_KEY_MOUSE_KEYBOARD,
    ACTION_COLOR_KEY_VARIABLES, ACTION_COLOR_KEY_WAIT,
};
use std::collections::HashMap;
use std::sync::RwLock;

static CUSTOM_ACTION_COLORS: RwLock<Option<HashMap<String, [u8; 4]>>> = RwLock::new(None);

fn action_color_key(action_type: &str) -> &'static str {
    taxonomy_color_key(action_type)
}

/// Sample action type used when previewing a category swatch.
pub fn sample_action_type_for_color_key(category_key: &str) -> &'static str {
    match category_key {
        ACTION_COLOR_KEY_MOUSE_KEYBOARD => "click",
        ACTION_COLOR_KEY_DETECTION => "imagesearch",
        ACTION_COLOR_KEY_VARIABLES => "setvariable",
        ACTION_COLOR_KEY_CONTROL_FLOW => "loop",
        ACTION_COLOR_KEY_MISCELLANEOUS => "focuswindow",
        ACTION_COLOR_KEY_WAIT => "wait",
        _ => "",
    }
}

/// Store a user-chosen color for a category key.
pub fn set_custom_action_color(category_key: &str, rgba: [u8; 4]) {
    let mut guard = CUSTOM_ACTION_COLORS.write().unwrap();
    let map = guard.get_or_insert_with(HashMap::new);
    map.insert(category_key.to_string(), rgba);
}

/// Remove a user override for a category key.
pub fn clear_custom_action_color(category_key: &str) {
    let mut guard = CUSTOM_ACTION_COLORS.write().unwrap();
    if let Some(map) = guard.as_mut() {
        map.remove(category_key);
    }
}

/// Remove every user override.
pub fn clear_all_custom_action_colors() {
    *CUSTOM_ACTION_COLORS.write().unwrap() = None;
}

/// Category pastel color, light/dark theme.
/// Uses a user override when one is set; otherwise the built-in pastel.
pub fn action_pastel_color(action_type: &str, is_dark: bool) -> [u8; 4] {
    let t = action_type.trim().to_ascii_lowercase();
    if t != "warning" {
        let key = action_color_key(&t);
        if let Some(c) = CUSTOM_ACTION_COLORS
            .read()
            .unwrap()
            .as_ref()
            .and_then(|m| m.get(key).copied())
        {
            return c;
        }
    }
    default_action_pastel_color(action_type, is_dark)
}

/// Built-in pastel, ignoring user overrides.
pub fn default_action_pastel_color(action_type: &str, is_dark: bool) -> [u8; 4] {
    let t = action_type.trim().to_ascii_lowercase();
    if t == "warning" {
        return if is_dark {
            [0x8A, 0x5A, 0x2A, 0xFF]
        } else {
            [0xF0, 0xC0, 0x6A, 0xFF]
        };
    }
    let key = action_color_key(&t);

    if is_dark {
        return match key {
            ACTION_COLOR_KEY_WAIT => [0x7B, 0x4E, 0x3E, 0xFF],
            ACTION_COLOR_KEY_MOUSE_KEYBOARD => [0x5E, 0x6B, 0x4A, 0xFF],
            ACTION_COLOR_KEY_DETECTION => [0x5A, 0x4A, 0x44, 0xFF],
            ACTION_COLOR_KEY_VARIABLES => [0x2A, 0x42, 0x54, 0xFF],
            ACTION_COLOR_KEY_CONTROL_FLOW => [0x3A, 0x5A, 0x58, 0xFF],
            ACTION_COLOR_KEY_MISCELLANEOUS => [0x8A, 0x45, 0x68, 0xFF],
            _ => [0x5C, 0x54, 0x49, 0xFF],
        };
    }
    match key {
        ACTION_COLOR_KEY_WAIT => [0xC9, 0x8D, 0x6A, 0xFF],
        ACTION_COLOR_KEY_MOUSE_KEYBOARD => [0xA1, 0xB0, 0x7A, 0xFF],
        ACTION_COLOR_KEY_DETECTION => [0xB4, 0x9A, 0x84, 0xFF],
        ACTION_COLOR_KEY_VARIABLES => [0x5E, 0x8F, 0xB0, 0xFF],
        ACTION_COLOR_KEY_CONTROL_FLOW => [0x7A, 0xB8, 0xB0, 0xFF],
        ACTION_COLOR_KEY_MISCELLANEOUS => [0xE0, 0x90, 0xB0, 0xFF],
        _ => [0xB2, 0xA4, 0x8E, 0xFF],
    }
}

/// Nested `${var}` chip fill.
pub fn nested_var_ref_color(is_dark: bool) -> [u8; 4] {
    if is_dark {
        [0x46, 0x62, 0x78, 0xFF]
    } else {
        [0x9E, 0xC4, 0xE3, 0xFF]
    }
}
